kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,273 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+      public_event_group_ids.emplace_back(global_eg_id);
+    }
+  }
+  // populate private_event_group_ids
+  if (eg_state->private_offset != 0) {
+    for (uint64_t i = eg_state->private_offset; i < std::min(eg_state->private_offset + kBatchSize, private_end + 1);
+         ++i) {
+      uint64_t global_eg_id =
+          getValueFromTagTable(client_id + kTagTableKeySeparator + concordUtils::toBigEndianStringBuffer(i));
+      private_event_group_ids.emplace_back(global_eg_id);
+    }
+  }
+  // populate event_group_id_batch with only public event group ids iff one of the following occurs -\
+  // 1. No private event groups exist in storage \
+  // 2. The first global event group id in private_event_group_ids is greater than the last global event group id public_event_group_ids\
+  // Vice versa is true for populate event_group_id_batch with only private event group ids \
+  // Note: each private_event_group_ids and public_event_group_ids hold a sorted list of global event group ids at all times
+  if (!public_event_group_ids.empty() &&
+      (private_event_group_ids.empty() || (public_event_group_ids.back() <= private_event_group_ids.front()))) {
+    for (auto public_event_group_id : public_event_group_ids) {
+      eg_state->event_group_id_batch.emplace_back(public_event_group_id);
+    }
+    ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
+    // increment the offset
+    eg_state->public_offset += eg_state->event_group_id_batch.size();
+    LOG_DEBUG(logger_,
+              "Updated public_offset: " << eg_state->public_offset
+                                        << " public_event_group_ids size: " << public_event_group_ids.size());
+  } else if (!private_event_group_ids.empty() &&
+             (public_event_group_ids.empty() || (private_event_group_ids.back() <= public_event_group_ids.front()))) {
+    for (auto private_event_group_id : private_event_group_ids) {
+      eg_state->event_group_id_batch.emplace_back(private_event_group_id);
+    }
+    ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
+    // increment the offset
+    eg_state->private_offset += eg_state->event_group_id_batch.size();
+    LOG_DEBUG(logger_,
+              "Updated private_offset: " << eg_state->private_offset
+                                         << " private_event_group_ids size: " << private_event_group_ids.size());
+  } else if (!public_event_group_ids.empty() && !private_event_group_ids.empty()) {
+    LOG_DEBUG(logger_, "Both public and private event groups found in the batch");
+    std::merge(public_event_group_ids.begin(),
+               public_event_group_ids.end(),
+               private_event_group_ids.begin(),
+               private_event_group_ids.end(),
+               std::back_inserter(eg_state->event_group_id_batch));
+    // We cannot return the full merge
+    // E.g.: a=[1,2,3,4,5] b=[6,7,8,9]
```

Comment:
In order to make this example more obvious, we should mention the batch size. E.g. set `kBatchSize=2` or similar.

Reply:
Updated.